### DISCUSSION TITLE
PowerShell with low-code capabilities

### DISCUSSION THREAD

**Discussion by anderson-joyle** [PowerShell/PowerShell#19806] [General]
I wondered if embedding [Power-Fx](https://github.com/microsoft/Power-Fx) into PowerShell would bring benefits. Having an expression like
`Get-ChildItem |
    Sort-Object -Property LastWriteTime, Name |
    Format-Table -Property LastWriteTime, Name`

simplified into something like
`ListFiles(Sort(LastWriteTime, Name),Columns(LastWriteTime, Name))`

**Comment by ThomasNieto** [comment]
What use cases are you looking for when adding those capabilities from Power-Fx? Is it to be more concise when writing commands? If so, you could use aliases in your example to trim down the command to as short or shorter than the example.

```powershell
gci | sort LastWriteTime, Name | ft LastWriteTime, Name
```

**Reply by anderson-joyle** [reply]
As @kilasuit already mentioned, its less the concise aspect and more the familiarity with an expression similar to Excel.

**Comment by MartinGC94** [comment]
I don't see how that can be considered "simplified". Keeping track of the parentheses seem more difficult than simply placing a pipe symbol after each command.  
You can already leave out the parameter names in PS, but IMO it only makes sense to do so when it's clear at a glance what the arguments are used for.  
I will give you that `ListFiles` is more beginner friendly than `Get-ChildItem` but that's not really a language problem, it could just as easily have been called `Get-Files` in PS. I don't think the `Columns` name is better than the PS equivalent, which would be `Select` and not `Format-Table.  
So, is it more simple than: `Get-Files | Sort LastWriteTime, Name | Select LastWriteTime, Name` ? I don't think so.

IMO the existing PowerShell syntax is already pretty user friendly for non-developers.

**Reply by kilasuit** [reply]
Get-ChildItem is used for more than just Files though as is used across different Providers, so includes Certs, Functions, Aliases, Active Directory Objects, Registry Items to name a few

**Reply by MartinGC94** [reply]
That is correct. I'm just saying from a beginners perspective, the generic name does make things a bit more difficult.

**Comment by mattcargile** [comment]
As another perspective on the _third-party_ module circuit, you might be interested in [`PipeScript`](https://github.com/StartAutomating/PipeScript) by [_StartAutomating_](https://github.com/StartAutomating). It can produce `powershell` in non-typical formats that transpile down to `powershell`.

Personally, I love the piping and aliasing already in the language. It is more core to a _shell_ in general and I wouldn't be keen to use a more _methods-based_ approach. And like others have said, I don't think this would be simpler. If anything, this approach could probably be implemented in _.NET_ or the like as a class.

**Comment by kilasuit** [comment]
I can see why coming from PowerFX & Excel (where this syntax is also used) it may feel more natural to use that syntax however as mentioned above each additional wrapping of parentheses (also known as round brackets) is harder to keep track of for many people (not all/most) but also harder for code reviewing and any automated complexity/maintainability metrics that may be in use in organisations.

As PowerFx is a nuget package you could pull this into your own PowerShell environment and use their methods today, however I don't think you'd get a great experience doing so.

**Reply by anderson-joyle** [reply]
I see your point, and I appreciate your comment. 👍

**Reply by kilasuit** [reply]
This `ListFiles(Sort(LastWriteTime, Name),Columns(LastWriteTime, Name))` would likely look like this if you were to pull in the methods from the PowerFX nuget packages

```powershell
[PowerFX]::ListFiles([PowerFx]::Sort(LastWriteTime, Name),[PowerFX]::Columns(LastWriteTime, Name))
```
Which makes it even harder to read but shows as an example of using .Net methods in PowerShell as PowerFX is built in C# 
What needs to be in `[<here>]::` is a guess btw & likely wrong depending on which of the packages you bring in.
